docs/tutorials/basics/basics_promotion.md

```diff
@@ -0,0 +1,130 @@
+has to make sure that a paused trial can be resumed later on, continuing
+training as if nothing happened in between. To this end, the *state* of the
+training job has to be checkpointed (i.e., stored into a file).
+
+Please have a look at [traincode_report_withcheckpointing.py](scripts/traincode_report_withcheckpointing.py).
+Checkpointing needs to do two things:
+* [1] A checkpoint has to be written at the end of each epoch. The precise
+  content of the checkpoint depends on the training script, but it has to
+  contain the epoch at which it was written.
+* [2] A checkpoint is to be loaded just before the start of the training loop.
+  If the checkpoint file is present and the state can be restored, the
+  training loop starts with the epoch `resume_from + 1`, where the checkpoint
+  was written at the end of epoch `resume_from`. Otherwise, `resume_from = 0`,
+  and the training loop starts from scratch.
+* [3] Checkpointing requires additional input arguments. You can add them by
+  hand or use `add_checkpointing_to_argparse`. The most important is the
+  local directory name where the checkpoint should be written or loaded from.
+  A checkpoint may consist of different files. If this argument is not
+  passed to the script, checkpointing is deactivated.
+
+Syne Tune provides some helper functions for checkpointing in
+[checkpoint.py](../../../benchmarks/checkpoint.py).
+* `checkpoint_model_at_rung_level(config, save_model_fn, epoch)` stores
+  a checkpoint at the end of epoch `epoch`. The main work is done by
+  `save_model_fn`.
+* `resume_from = resume_from_checkpointed_model(config, load_model_fn)` loads a
+  checkpoint, and returns its epoch if successful. Otherwise, 0 is returned.
+  Again, `load_model_fn` does the main work.
+* `pytorch_load_save_functions`: If you use *PyTorch*, this is providing
+  `save_model_fn`, `load_model_fn` that should work for you. In
+  `state_dict_objects`, you pass a dict of PyTorch objects with a mutable
+  state (look for `load_state_dict`, `state_dict` methods). Make sure to include
+  all relevant objects (model, algorithm, learning rate scheduler). Optionally,
+  `mutable_state` contains additional elementary variables.
+
+Note that while checkpoints are written at the end of each epoch, the most
+recent one overwrites previous ones. In fact, for the purpose of pause and
+resume, checkpoints have to be written only at rung levels, because trials
+can only be paused there. Selective checkpointing could be supported by passing
+the rung levels to the training script, but this is currently not done in
+Syne Tune.
+
+
+## Launcher Script for promotion-based ASHA and MOBSTER
+
+Once a training script supports checkpointing, the promotion variant of ASHA
+MOBSTER are obtained by selecting `type='promotion'`, see
+[launch_asha_promotion.py](scripts/launch_asha_promotion.py) and
+[launch_mobster_promotion/py](scripts/launch_mobster_promotion.py).
```

Comment:
Having duplicate examples for all options will be hard to maintain (they will easily break with code changes). 

I wonder about options such as:
* 1 linking examples that already exist in `examples/`
* 2 linking example that are in a specific branch (this way they wont break)
* 3 giving only the sniplet to create the scheduler (given that the example only changes the scheduler option)

Creating a specific branch and linking from there (option 3) seems to me the easiest to make sure the examples still run.

Reply:
I'd like to allow people to run these scripts, that is why they are complete.

We still don't have a solution for integration testing. Once we have one (with some engineering support hopefully), we could include the main tutorials by say running each launcher script until a time cutoff and check whether it does not fail.

As long as we don't have a go-to page for docs, I'd like to make sure the docs are part of the main branch.

It is just one of these things. Beyond all clever mechanisms, once in a while we need to do a sprint going over docs nd tutorials (the most important ones) and check whether they still work. We'll also get feedback from users. It is not terrible for an OS project that not every tutorial works all the time, given we fix things when told about.

Reply:
One possibility, which we could do in another PR, is to be more economical and create an "uber" launch script with options at the top.

But I am not sure. This would again be complicated, no?

Reply:
I think that if we provide full snippets then we have to make sure that they work as it is a very bad experience if they don't (and having a sprint consisting of manually testing the tutorial is not a great option).

Can we put the tutorial in main but the snippets in a specific branch to ensure the examples always work?

Reply:
No because then if the user clicks on the code, they will get an error there, because they don't have the right branch.

Reply:
To be clear: In the mid term, we need to find a better solution to docs / tutorials, such as a readthedocs.io page. There, we could try and build tutorials for a special branch, if that is possible.

As for now, we should make sure that whenever we release a new version to PyPI, all docs and tutorials are checked. Checking a tutorial's code is no different from an integration test.

Reply:
In fact, we should start to define a process for releases to PyPI.

Reply:
One idea would be to adopt readthedocs.io and build that docs page from the branch of the latest release, maybe with a "risky" option to build it from the current main.

Do you know anybody who could help with setting up readthedocs? I skimmed the docs, we need to also adopt something called sphinx, which I think is a specific way to annotate code.

Reply:
> No because then if the user clicks on the code, they will get an error there, because they don't have the right branch.

We could very well link both the branch and explain clearly how to run the tutorial examples, eg to indicate that the snippets can be run by doing this:

```bash 
git checkout tutorial
python tutorial/example_foo.py
```

This will make sure that the tutorial doc is on main and that the code always run.

If this is not an option lets ship as it is, however I do not think we should put all tutorial examples in the integration tests as they duplicate a lot of examples already running and the CI would run significantly longer.

Reply:
OK, what I could do is add some instructions to the start of the tutorial, saying something like:
If you want to run the code examples linked in this tutorial, be sure to use this release (with instructions how to get it).

Note that the code being part of this tutorial is not meant to be run as part of the tutorial. But the Fair HPO tutorial would be a different example.

Reply:
Ok, can you add such comment in all code examples too? (since people may find them without passing from the tutorial doc)

Reply:
OK, done that.
We'll get to a better solution, where people do not have to clone the repo.